src/hera/workflows/v5/_context.py

```diff
@@ -17,11 +17,15 @@ def enter(self, p: Subbable) -> None:
         self._pieces.append(p)
 
     def exit(self) -> None:
-        self._pieces.pop()
+        if len(self._pieces) == 1:
+            return  # only the workflow is in the context
 
-    def add_sub_node(self, node: _SubNodeMixin) -> None:
-        if self._pieces:
-            self._pieces[-1]._add_sub(node)
+        popped = self._pieces.pop()
+        main = self._pieces[0]
```

Comment:
I think we also need to make sure we do the _build_* calls for each sub node from the parent container when the build call is made to the parent container, otherwise we might miss changes.